### ISSUE TITLE
Add first-class concept of topics

### ISSUE THREAD

**Issue by fabxc** [oklog/oklog#113]
Oklog does provide no first-class concept for logical separation of log streams (think Kafka topics). However, it can be easily implemented by prefixing log records by a topic key. If one only wants to query a specific topic, records must be filtered along the given prefix.

This can work well for some use cases but lead to unacceptable read amplification for others.

A first class notion of topics can address this problem. If data on disk is grouped by topic, we can filter out all irrelevant topics without read overhead.  
Topics should be an opaque key with only basic selection properties (equal, in-set, has-prefix, regex(?)) and not be abused for advanced record indexing tasks. An intended use case would be "topic per Kubernetes namespace" or "topic per service".

Making topics a first-class concept likely requires touching virtually all APIs of oklog.

The disk layout will need to change to group records by topic on disk.
If topic cardinality is high, this may notably shift performance characteristics. 1,000-10,000 topics per oklog cluster sounds like a reasonable upper bound.

**Comment by peterbourgon** [comment]
My understanding is that topic would need to be first-order parameter of any query, initially just selecting specific topic, maybe eventually allowing querying multiple topics at once, but never anything more sophisticated than that, is this correct?

**Comment by fabxc** [comment]
Yes, fully agree. Simple equal matching is the first thing we want. There'll be use cases for cross-topic queries rather quickly I assume but that can be figured out along the way.

**Comment by peterbourgon** [comment]
What about this? (Not necessarily suggesting it, just to provoke discussion.)

- If the ingesters are deployed in `-topic-mode none` mode, then all records are assumed to be part of the same default unnamed/empty topic.
- If the ingesters are deployed in `-topic-mode static -topic xxx` mode, then all records are assumed to be part of the same `xxx` topic. (`-topic-mode none` is a degenerate case of this mode.)
- If the ingesters are deployed in `-topic-mode dynamic` mode, then the first whitespace separated token of each record is interpreted as the topic.
- If the ingesters are deployed in `-topic-mode connection` mode, then the first whitespace separated token on the first line sent on each new connection is interpreted as the fixed topic for all subsequent records on that connection.

**Comment by peterbourgon** [comment]
The problem here is if we mandate use of `oklog forwarder` on the other side of the `ingester`, I'm convinced we make things very annoying for many people. I'm imagining that lots of environments have fluentd or logspout or some custom file-oriented thing baked in as the log shipper; newline-delimited records seem like the least-common-denominator. But I'm open to being persuaded this isn't the case...

**Comment by fabxc** [comment]
Different topic modes definitely makes. Especially since there are existing users of oklog who'd break otherwise.  
The `connection` option sounds potentially useful. Though I'm not sure the standard log forwarders out there support setting this up. Probably something to keep in mind and wait until there's a use case for it.  
I could imagine it just not being worth the small performance benefit over just using `dynamic` and adding a static prefix to each record.

Totally agree mandating `oklog forwarder` being a usability problem – and unless we want to switch to a fully structured API, which seems unnecessary, there's no strong reason for it.

An additional structured ingester API may make sense if we want to allow arbitrary records at some point. But I see it in a distant future and don't want to push this topic more than I'm doing already ;0

**Comment by fabxc** [comment]
I tried to give the ingester side of things a stab today. The topic assignment (`none`, `static`, `dynamic`) can be integrated nicely without much trouble.
Splitting segments into directory per topic is obviously more involved. It also raised some efficiency questions, most importantly: segments will mostly complete through timeout and be tiny. That, like everything, can be addressed but probably won't work with the nice and simple segment pull store nodes have right now.

It seems feasible to just leave ingesters as they are (except for prefixing static/default topic) and instead split and scatter pulled segments at the store side. That adds a processing pass over incoming data, in return the ingesters and segment pull API can stay conceptually simple.

**Comment by fabxc** [comment]
Maybe simpler to actually just have `-topic-mode={dynamic,static}` and `-topic` default to `_default` or something. That's what we'd do internally anyway and masking that fact may not be too helpful to the user in the end.

**Comment by peterbourgon** [comment]
For context can you describe the use-case where mode=dynamic is most useful? I'm struggling a bit because my assumption is that a single connection and forwarder would map to a single service, which would necessarily imply a single topic — but there's probably something I'm overlooking.

**Comment by fabxc** [comment]
Suppose we map an oklog topic to a Kubernetes namespace/service. Namespaces/services are just a logical concept but pods for any namespace/service can get scheduled onto any node (in general). The log forwarder (from my understanding) would simply forward logs for all containers it can find on that node through a single connection. Thus the "physical" connection could transport logs for arbitrary topics of the cluster.

**Comment by tsenart** [comment]
> Thus the "physical" connection could transport logs for arbitrary topics of the cluster.

Essentially, a form of multiplexing.
